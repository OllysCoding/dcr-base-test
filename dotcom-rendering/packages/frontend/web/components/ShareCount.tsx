import React from 'react';
import { css } from 'emotion';
import ShareIcon from '@frontend/static/icons/share.svg';
import {
    from,
    wide,
    leftCol,
    textSans,
    palette,
} from '@guardian/src-foundations';
import { integerCommas } from '@frontend/lib/formatters';
import { useApi } from '@frontend/web/components/lib/api';

const shareCount = css`
    ${textSans({ level: 3 })};
    font-weight: bold;
    color: ${palette.neutral[46]};

    ${leftCol} {
        border-top: 1px solid ${palette.neutral[86]};
        width: 100%;
        padding-top: 6px;
    }

    ${wide} {
        flex: 1;
        border: 0;
        padding-top: 0;
        text-align: right;
    }
`;

const shareCountContainer = css`
    ${leftCol} {
        display: inline-block;
    }
`;

const shareCountHeader = css`
    position: relative;
    height: 15px;
    margin: 0;
`;

const shareCountIcon = css`
    position: absolute;
    top: 0;
    right: 0;
    fill: ${palette.neutral[46]};
`;

const countFull = css`
    display: block;

    ${from.leftCol.until.wide} {
        display: none;
    }
`;

const countShort = css`
    display: none;

    ${from.leftCol.until.wide} {
        display: block;
    }
`;

interface Props {
    config: ConfigType;
    pageId: string;
}

interface ShareCountType {
    path: string;
    share_count: number;
    refreshStatus: boolean;
}
function buildUrl(ajaxUrl: string, pageId: string) {
    return `${ajaxUrl}/sharecount/${pageId}.json`;
}

export const ShareCount = ({ config, pageId }: Props) => {
    const url = buildUrl(config.ajaxUrl, pageId);
    const { data, error } = useApi<ShareCountType>(url);

    if (error) {
        window.guardian.modules.raven.reportError(
            error,
            {
                feature: 'share-count',
            },
            false,
        );

        return null;
    }

    if (!data || !data.share_count) {
        return null;
    }

    const displayCount = parseInt(data.share_count.toFixed(0), 10);
    const formattedDisplayCount = integerCommas(displayCount);
    const shortDisplayCount =
        displayCount > 10000
            ? `${Math.round(displayCount / 1000)}k`
            : displayCount;

    return (
        <div className={shareCount} aria-label={`${shortDisplayCount} Shares`}>
            <div className={shareCountContainer}>
                <div className={shareCountHeader}>
                    <ShareIcon className={shareCountIcon} />
                </div>
                <div
                    data-testid={'countFull'}
                    className={countFull}
                    aria-hidden="true"
                >
                    {formattedDisplayCount}
                </div>
                <div
                    data-testid={'countShort'}
                    className={countShort}
                    aria-hidden="true"
                >
                    {shortDisplayCount}
                </div>
            </div>
        </div>
    );
};
