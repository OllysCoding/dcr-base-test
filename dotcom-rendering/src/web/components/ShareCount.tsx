import React from 'react';
import { css } from 'emotion';
import ShareIcon from '@frontend/static/icons/share.svg';
import { palette } from '@guardian/src-foundations';
import { textSans } from '@guardian/src-foundations/typography';
import { from, between } from '@guardian/src-foundations/mq';
import { integerCommas } from '@root/src/lib/formatters';
import { useApi } from '@root/src/web/components/lib/api';

type Props = {
    ajaxUrl: string;
    pageId: string;
};

type ShareCountType = {
    path: string;
    share_count: number;
    refreshStatus: boolean;
};

const shareCount = css`
    ${textSans.medium()};
    font-weight: bold;
    color: ${palette.neutral[46]};

    ${from.leftCol} {
        border-top: 1px solid ${palette.neutral[86]};
        width: 100%;
        padding-top: 6px;
    }

    ${from.wide} {
        flex: 1;
        border: 0;
        padding-top: 0;
        text-align: right;
    }
`;

const shareCountContainer = css`
    ${from.leftCol} {
        display: inline-block;
    }
`;

const shareCountHeader = css`
    position: relative;
    height: 15px;
    margin: 0;
`;

const shareCountIcon = css`
    position: absolute;
    top: 0;
    right: 0;
    fill: ${palette.neutral[46]};
`;

const countFull = css`
    display: block;

    ${between.leftCol.and.wide} {
        display: none;
    }
`;

const countShort = css`
    display: none;

    ${between.leftCol.and.wide} {
        display: block;
    }
`;

export const ShareCount = ({ ajaxUrl, pageId }: Props) => {
    const url = `${ajaxUrl}/sharecount/${pageId}.json`;
    const { data, error } = useApi<ShareCountType>(url);

    if (error) {
        window.guardian.modules.sentry.reportError(error, 'share-count');

        return null;
    }

    if (!data || !data.share_count) {
        return null;
    }

    const countAsInteger = parseInt(data.share_count.toFixed(0), 10);
    const displayCountLong = integerCommas(countAsInteger);
    const displayCountShort =
        countAsInteger > 10000
            ? `${Math.round(countAsInteger / 1000)}k`
            : countAsInteger;

    return (
        <div
            className={shareCount}
            aria-label={`${displayCountShort} Shares`}
            data-cy="share-count"
        >
            <div className={shareCountContainer}>
                <div className={shareCountHeader}>
                    <ShareIcon className={shareCountIcon} />
                </div>
                <div
                    data-testid="countFull"
                    className={countFull}
                    aria-hidden="true"
                >
                    {displayCountLong}
                </div>
                <div
                    data-testid="countShort"
                    className={countShort}
                    aria-hidden="true"
                >
                    {displayCountShort}
                </div>
            </div>
        </div>
    );
};
