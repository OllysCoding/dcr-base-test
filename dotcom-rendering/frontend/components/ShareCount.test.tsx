import React from 'react';
import { shallow } from 'enzyme';
import { ShareCount } from './ShareCount';
import { CAPI } from '../../test/fixtures/CAPI';

const fetchShareCount: (shareCount: number) => Promise<{}> = shareCount =>
    new Promise(resolve => {
        resolve({
            ok: true,
            json: () => ({
                share_count: shareCount,
            }),
        });
    });

describe('ShareCount', () => {
    const globalAny: any = global;
    const config: ConfigType = {
        ajaxUrl: 'https://api.nextgen.guardianapps.co.uk',
    };

    afterEach(() => {
        const { pageId } = CAPI;
        const { ajaxUrl } = config;

        expect(globalAny.fetch).toHaveBeenCalledWith(
            `${ajaxUrl}/sharecount/${pageId}.json`,
        );
    });

    describe('snapshots', () => {
        it('It should not render if state.shareCount is falsy', () => {
            const fakeFetch = fetchShareCount(0);

            globalAny.fetch = jest.fn().mockImplementation(() => fakeFetch);

            const component = shallow(
                <ShareCount config={config} CAPI={CAPI} />,
            );

            return fakeFetch
                .then(() => {
                    component.update();
                })
                .then(() => {
                    expect(component).toMatchSnapshot();
                });
        });

        it('It should render if state.shareCount is truthy', () => {
            const fakeFetch = fetchShareCount(100);

            globalAny.fetch = jest.fn().mockImplementation(() => fakeFetch);

            const component = shallow(
                <ShareCount config={config} CAPI={CAPI} />,
            );

            return fakeFetch
                .then(() => {
                    component.update();
                })
                .then(() => {
                    expect(component).toMatchSnapshot();
                });
        });

        it('It should format long shareCount correctly', () => {
            const fakeFetch = fetchShareCount(25000);

            globalAny.fetch = jest.fn().mockImplementation(() => fakeFetch);

            const component = shallow(
                <ShareCount config={config} CAPI={CAPI} />,
            );

            return fakeFetch
                .then(() => {
                    component.update();
                })
                .then(() => {
                    expect(component).toMatchSnapshot();
                });
        });
    });

    it('It should render null if state.shareCount is falsy', () => {
        const fakeFetch = fetchShareCount(0);

        globalAny.fetch = jest.fn().mockImplementation(() => fakeFetch);

        const component = shallow(<ShareCount config={config} CAPI={CAPI} />);

        return fakeFetch
            .then(() => {
                component.update();
            })
            .then(() => {
                expect(component.type()).toBeNull();
                expect(component.state('shareCount')).toBe(0);
            });
    });

    it('It should render if state.shareCount is truthy', () => {
        const fakeFetch = fetchShareCount(100);

        globalAny.fetch = jest.fn().mockImplementation(() => fakeFetch);

        const component = shallow(<ShareCount config={config} CAPI={CAPI} />);

        return fakeFetch
            .then(() => {
                component.update();
            })
            .then(() => {
                expect(component.type()).not.toBeNull();
                expect(component.state('shareCount')).toBe(100);
                expect(component.find('.js-countFull').text()).toBe('100');
                expect(component.find('.js-countShort').text()).toBe('100');
            });
    });

    it('It should format long shareCount correctly', () => {
        const fakeFetch = fetchShareCount(25000);

        globalAny.fetch = jest.fn().mockImplementation(() => fakeFetch);

        const component = shallow(<ShareCount config={config} CAPI={CAPI} />);

        return fakeFetch
            .then(() => {
                component.update();
            })
            .then(() => {
                expect(component.type()).not.toBeNull();
                expect(component.state('shareCount')).toBe(25000);
                expect(component.find('.js-countFull').text()).toBe('25,000');
                expect(component.find('.js-countShort').text()).toBe('25k');
            });
    });
});
