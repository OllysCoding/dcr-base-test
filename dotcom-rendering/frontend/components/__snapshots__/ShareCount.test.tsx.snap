// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareCount snapshots It should format long shareCount correctly 1`] = `
.emotion-6 {
  font-size: 18px;
  line-height: 18px;
  font-family: GuardianTextSans,Helvetica Neue,Helvetica,Arial,Lucida Grande,sans-serif;
  font-weight: bold;
  color: #767676;
}

@media (min-width:1140px) {
  .emotion-6 {
    border-top: 1px solid #dcdcdc;
    width: 100%;
    padding-top: 6px;
  }
}

@media (min-width:1300px) {
  .emotion-6 {
    -webkit-flex: 1;
    -ms-flex: 1;
    flex: 1;
    border: 0;
    padding-top: 0;
    text-align: right;
  }
}

@media (min-width:1140px) {
  .emotion-5 {
    display: inline-block;
  }
}

.emotion-2 {
  position: relative;
  height: 15px;
  margin: 0;
}

.emotion-0 {
  position: absolute;
  top: 0;
  right: 0;
  fill: #767676;
}

.emotion-3 {
  display: block;
}

@media (min-width:1140px) and (max-width:1299px) {
  .emotion-3 {
    display: none;
  }
}

.emotion-4 {
  display: none;
}

@media (min-width:1140px) and (max-width:1299px) {
  .emotion-4 {
    display: block;
  }
}

.emotion-1 {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  overflow: hidden;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  white-space: nowrap;
  border: 0;
}

<div
  className="emotion-6"
>
  <div
    className="emotion-5"
  >
    <h3
      className="emotion-2"
    >
      <SVG
        className="emotion-0"
      />
      <span
        className="emotion-1"
      >
        Shares
      </span>
    </h3>
    <div
      className="js-countFull emotion-3"
    >
      25,000
    </div>
    <div
      className="js-countShort emotion-4"
    >
      25k
    </div>
  </div>
</div>
`;

exports[`ShareCount snapshots It should not render if state.shareCount is falsy 1`] = `""`;

exports[`ShareCount snapshots It should render if state.shareCount is truthy 1`] = `
.emotion-6 {
  font-size: 18px;
  line-height: 18px;
  font-family: GuardianTextSans,Helvetica Neue,Helvetica,Arial,Lucida Grande,sans-serif;
  font-weight: bold;
  color: #767676;
}

@media (min-width:1140px) {
  .emotion-6 {
    border-top: 1px solid #dcdcdc;
    width: 100%;
    padding-top: 6px;
  }
}

@media (min-width:1300px) {
  .emotion-6 {
    -webkit-flex: 1;
    -ms-flex: 1;
    flex: 1;
    border: 0;
    padding-top: 0;
    text-align: right;
  }
}

@media (min-width:1140px) {
  .emotion-5 {
    display: inline-block;
  }
}

.emotion-2 {
  position: relative;
  height: 15px;
  margin: 0;
}

.emotion-0 {
  position: absolute;
  top: 0;
  right: 0;
  fill: #767676;
}

.emotion-3 {
  display: block;
}

@media (min-width:1140px) and (max-width:1299px) {
  .emotion-3 {
    display: none;
  }
}

.emotion-4 {
  display: none;
}

@media (min-width:1140px) and (max-width:1299px) {
  .emotion-4 {
    display: block;
  }
}

.emotion-1 {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  overflow: hidden;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  white-space: nowrap;
  border: 0;
}

<div
  className="emotion-6"
>
  <div
    className="emotion-5"
  >
    <h3
      className="emotion-2"
    >
      <SVG
        className="emotion-0"
      />
      <span
        className="emotion-1"
      >
        Shares
      </span>
    </h3>
    <div
      className="js-countFull emotion-3"
    >
      100
    </div>
    <div
      className="js-countShort emotion-4"
    >
      100
    </div>
  </div>
</div>
`;
